### PR TITLE
[Multi-task Learning] Add support for multi-task learning

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
 #789

*Description of changes:*
Add support for multi-task learning. Users can define multiple tasks in the same training loop. A task can be a node classification, node regression, edge classification, edge regression or link prediction task. For each node classification or node regression task, it should be defined on a single node type with one label field. But users can define multiple node classification or regression tasks on the same node type. For each edge classification or node regression task, it should be defined on a single edge type with one label field. But users can define multiple edge classification or regression tasks on the same edge type. For link prediction, users can define prediction targets on multiple edge types.

### Graph construction
Update GraphStorm input config parsing to support multi-task learning. Allow user to specify multiple training tasks for a training job through yaml file. By providing the `multi_task_learning` configurations in the yaml file, users can define multiple training tasks. The following config defines two training tasks, one for node classification and one for edge classification.

```
---
version: 1.0
gsf:
  basic:
    ...
  ...
  multi_task_learning:
    - node_classification:
      target_ntype: "movie"
      label_field: "label"
      mask_fields:
        - "train_mask_field_nc"
        - "val_mask_field_nc"
        - "test_mask_field_nc"
      task_weight: 1.0
    - edge_classification:
      target_etype:
        - "user,rating,movie"
      label_field: "rate"
      mask_fields:
        - "train_mask_field_ec"
        - "val_mask_field_ec"
        - "test_mask_field_ec"
      task_weight: 0.5 # weight of the task
```
Task specific hyperparameters in multi-task learning are same as thoses in single task learning, except that two new configs are required, i.e., mask_fields and task_weight. The mask_fields provides the training, validation and test masks for the task and the task_weight gives its loss weight.

### DataLoader for multi-task learning
Add GSgnnMultiTaskDataLoader to support multi-task learning. 

When initializing a GSgnnMultiTaskDataLoader, users need to provide two
inputs: 1) a list of config.TaskInfo objects recording the information
of each task and 2) a list of dataloaders corresponding to each training
task.

During training for each iteration, GSgnnMultiTaskDataLoader will
iteratively call each task-dataloader to generate a mini-batch and
finally return a list of mini-batches to the trainer.

The length of the dataloader (number of batches for an epoch) is
determined by the largest task in the GSgnnMultiTaskDataLoader.
#834 

### Evaluator for multi-task learning

GSgnnMultiTaskEvaluator accepts a set of Evaluators, in the format of
dict ({task_id: Evaluator, ...}) as input to initialize the multi-task
evaluator.

When doing evaluation, it accepts three arguements val_results,
test_results and total_iters. The val_results and test_results will be
dicts in the format of {task_id_0: reslut, task_id_1: result}. The
GSgnnMultiTaskEvaluator will call task specify evaluators for each task
to compute the evaluation scores.
#837 

### Refactor graphstorm.model for multi-task learning

As multi-task learning trainer will invoke edge_mini_batch_predict,
lp_mini_batch_predict and node_mini_batch_predict when conducting
evaluation or testing, refactor the code to allow the functions to work
with different decoders.
#843 

### Add GSgnnMultiTaskSharedEncoderModel
GSgnnMultiTaskSharedEncoderModel allows multiple tasks to share the same
GNN encoder but have separate decoders for each task.
#855 

### Add Multi-task entrypoint
#849 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by zhjwy9343** [comment]
In this PR description, may add the limitation of `one task per one node/edge type` in the current multi-task implementation.

**Comment by classicsong** [comment]
> In this PR description, may add the limitation of `one task per one node/edge type` in the current multi-task implementation.

Updated.

File: training_scripts/gsgnn_mt/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+python3 -m graphstorm.gconstruct.construct_graph \
+	--conf-file $GS_HOME/tests/end2end-tests/data_gen/movielens_multi_task.json \
+	--num-processes 1 \
+	--output-dir movielen_100k_multi_task_train_val_1p_4t \
+	--graph-name movie-lens-100k \
+	--add-reverse-edges
+```
+
+## Run the example
+```
```
Comment (zhjwy9343) [inline training_scripts/gsgnn_mt/README.md:17]: Are we missing the command here?

File: training_scripts/gsgnn_mt/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+python3 -m graphstorm.run.gs_multi_task_learning \
+	--workspace $GS_HOME/training_scripts/gsgnn_mt  \
+	--num-trainers 4 \
```
Comment (zhjwy9343) [inline training_scripts/gsgnn_mt/README.md:20]: Because here is 1 partition, we can change or remove some unnecessary arguments, include:
`--num-trainer 1`, use 1 works for both CPUs and one GPU instance.
Remove:
`--num-samplers 0`
`--ip-config ip_list.txt`
`--ssh-port 2222`
`--logging-file /tmp/train_log.txt`
`--logging-level debug`

**Review by zhjwy9343 — APPROVED** [review]
Except for the last command, others LGTM
